### PR TITLE
Block compaction if the partition has TtlUpdate to repair

### PR DESCRIPTION
**Pull request by JingQianCloud**
If there is TtlUpdate need to be fixed, stop the compaction on that partition.

### CODE REVIEW

File: ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java [hunk 1]
```diff
@@ -248,6 +281,16 @@ private List<RequestInfo> getAmbryRequest(long partitionId) throws Exception {
...
   }
 
+  /**
+   * Get partitions which have RepairRequests to fix
+   * @return list of partition ids.
+   */
+  Set<Long> getPartitionIdsNeedRepair() throws SQLException {
+    // each server has 12 * 23 partition. right now search in one single query.
+    Set<Long> partitionsNeedRepair = db.getPartitionsNeedRepair(nodeId.getHostname(), nodeId.getPort(), partitionIds);
+    return partitionsNeedRepair;
```
Comment (justinlin-linkedin) [inline ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java:291]: nit: just return the set, we don't need to assign it to another variable.

File: ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java [hunk 2]
```diff
@@ -123,6 +132,30 @@ public void run() {
...
+              partitionsCompactionDisabled.add(partition);
+              PartitionId partitionId = clusterMap.getPartitionIdByName(Long.toString(partition));
+              storageManager.controlCompactionForBlobStore(partitionId, false);
+              logger.info("RepairRequests Sender: disable compaction on {} {}", nodeId, partitionId);
+            }
+          }
+
+          /*
+           * handle the RepairRequests
+           */
```
Comment (justinlin-linkedin) [inline ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java:158]: nit: can we create a new method for this logic?

File: ambry-mysql/src/main/java/com/github/ambry/repair/MysqlRepairRequestsDb.java [hunk 3]
```diff
@@ -248,6 +264,40 @@ public List<RepairRequestRecord> getRepairRequestsExcludingHost(long partitionId
...
+   * @param sourceHostName the host name of the source replica
+   * @param sourceHostPort the host port of the source replica
+   * @param partitions the partitions to check
+   * @return the partitions which have TtlUpdate requests to repair.
+   */
+  @Override
+  public Set<Long> getPartitionsNeedRepair(String sourceHostName, int sourceHostPort, List<Long> partitions)
+      throws SQLException {
+    String partitionsStr = partitions.stream().map(n -> n.toString()).collect(Collectors.joining(","));
+    String query = String.format(GET_PARTITIONS_QUERY_EXCLUDE_SOURCE_REPLICA_TEMPLATE + " (%s) ", partitionsStr);
```
[the remark this answers is not in the document]
Comment (JingQianCloud) [reply ambry-mysql/src/main/java/com/github/ambry/repair/MysqlRepairRequestsDb.java:291]: Unfortunately we cannot do that way. If we setString to the ?, the in clause is a string like IN ('0,2,4'). It was not the integer array as we hope. I tested it and it didn't work.

https://mkyong.com/jdbc/jdbc-preparedstatement-sql-in-condition/
https://stackoverflow.com/questions/25674226/how-to-pass-a-list-of-integer-value-with-preparedstatement

File: ambry-mysql/src/main/java/com/github/ambry/repair/MysqlRepairRequestsDb.java [hunk 4]
```diff
@@ -248,6 +264,40 @@ public List<RepairRequestRecord> getRepairRequestsExcludingHost(long partitionId
...
+          Set<Long> result = new HashSet<>();
+          while (resultSet.next()) {
+            Long p = resultSet.getLong(1);
+            result.add(p);
+          }
+          return result;
+        }
+      }
+    } catch (SQLException e) {
+      metrics.repairDbErrorGetCount.inc();
```
[the remark this answers is not in the document]
Comment (JingQianCloud) [reply ambry-mysql/src/main/java/com/github/ambry/repair/MysqlRepairRequestsDb.java]: Added the metrics

File: ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java (hunk 1 above)
Comment (JingQianCloud) [reply ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java]: Changed

File: ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java (hunk 2 above)
Comment (JingQianCloud) [reply ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java]: Wrap to one function.

**Review by SophieGuo410 — APPROVED** [review]

**Review by justinlin-linkedin — APPROVED** [review]

**Pull request merged**